esp-hal/src/parl_io.rs

```diff
@@ -1733,73 +1795,51 @@ pub mod asynch {
             cx: &mut core::task::Context<'_>,
         ) -> Poll<Self::Output> {
             TX_WAKER.register(cx.waker());
```

Comment:
You can move the TX registration to the `else` branch, just above `listen_tx_done`.